trafficFlow/carFollowingModel/drivers/baseDriver.py

```diff
@@ -34,22 +37,243 @@ class BaseDriver:
-        self.successor_right = None
         self.predecessor_left = None
+        self.successor_right = None
         self.successor_left = None
 
         self.object_in_visualization = None
 
+    def __str__(self):
+        return self.label
+
+    def get_distance_to_predecessor(self):
+        if self.predecessor and not self.predecessor == self:
+            return self.lane.road.get_distance(self.predecessor, self) - self.length
+        else:
+            return self.lane.full_length
+
+    def get_distance_to_successor(self):
+        if self.successor and not self.successor == self:
+            return self.lane.road.get_distance(self, self.successor) - self.successor.length
+        else:
+            return self.lane.full_length
+
+    def get_distance_to_predecessor_right(self):
+        if self.predecessor_right:
+            return self.lane.road.get_distance(self.predecessor_right, self) - self.length
+        else:
+            return self.lane.full_length
+
+    def get_distance_to_successor_right(self):
+        if self.successor_right:
+            return self.lane.road.get_distance(self, self.successor_right) - self.successor_right.length
+        else:
+            return self.lane.full_length
+
+    def get_distance_to_predecessor_left(self):
+        if self.predecessor_left:
+            return self.lane.road.get_distance(self.predecessor_left, self) - self.length
+        else:
+            return self.lane.full_length
+
+    def get_distance_to_successor_left(self):
+        if self.successor_left:
+            return self.lane.road.get_distance(self, self.successor_left) - self.successor_left.length
+        else:
+            return self.lane.full_length
+
+    def get_speed_difference_to_predecessor(self):
+        if self.predecessor and not self.predecessor == self:
+            return self.velocity - self.predecessor.velocity
+        else:
+            return -self.max_speed
+
+    def get_speed_difference_to_successor(self):
+        if self.successor and not self.successor == self:
+            return self.successor.velocity - self.velocity
+        else:
+            return math.inf
+
+    def get_speed_difference_to_predecessor_right(self):
+        if self.predecessor_right:
+            return self.velocity - self.predecessor_right.velocity
+        else:
+            return -math.inf
+
+    def get_speed_difference_to_successor_right(self):
+        if self.successor_right:
+            return self.successor_right.velocity - self.velocity
+        else:
+            return math.inf
+
+    def get_speed_difference_to_predecessor_left(self):
+        if self.predecessor_left:
+            return self.velocity - self.predecessor_left.velocity
+        else:
+            return -math.inf
+
+    def get_speed_difference_to_successor_left(self):
+        if self.successor_left:
+            return self.successor_left.velocity - self.velocity
+        else:
+            return math.inf
```

Comment:
This is a lot of code duplication, you should write a single ``get_distance`` and ``get_speed_difference`` function that takes a vehicle and pass ``self.predecessor_left``, ``self.successor_right`` etc. to these functions.

Reply:
If the implementation is changed such that  the ``vehicles`` list is ordered, the driver doesn't have to keep track of any neighbors. Instead, all required neighbors should be provided to the ``vehicle`` by the ``lane``.